File: libint-sys/shim/src/basis.cc
#include "libint-sys/shim/include/basis.h"
#include <memory>

namespace libint2 {
  std::unique_ptr<BasisSet> basis(rust::Str name, const Atom* const* atoms, size_t n) {
    // convert rust &str to std::string
    std::string s(name.data(), name.size());

    // convert pointer array of `Atom`s to std::vector<Atom>
    std::vector<Atom> tmp;
    tmp.reserve(n);
    for (size_t i = 0; i < n; ++i) {
        const libint2::Atom* a = atoms[i];
        tmp.push_back(*a);
    }

    // construct `BasisSet` and return
    return std::make_unique<BasisSet>(s, tmp);
  }

  std::size_t nshells(const BasisSet& basis) {
    return basis.size();
  }

  void set_pure(BasisSet& basis, bool solid) {
    basis.set_pure(solid);
  }

  std::unique_ptr<Shell> at(const BasisSet& basis, std::size_t i) {
    return std::make_unique<Shell>(basis[i]);
  }

  std::size_t nbf(const BasisSet& basis) {
    return basis.nbf();
  }

  std::size_t max_nprim(const BasisSet& basis) {
    return basis.max_nprim();
  }

  std::size_t max_l(const BasisSet& basis) {
    return basis.max_l();
  }
}

File: libint-sys/src/bridge/basis.rs
use cxx::CxxVector;

#[cxx::bridge(namespace = "libint2")]
pub mod ffi {

    unsafe extern "C++" {
        include!("libint-sys/shim/include/basis.h");

        type BasisSet;
        type Atom = crate::atom::Atom;
        type Shell = crate::shell::Shell;

        /// # Safety
        ///
        /// Make sure that `atoms` is not constructed with dangling pointers.
        /// This leads to C++ interpreting the pointers as garbage.
        #[must_use]
        unsafe fn basis(name: &str, atoms: *const *const Atom, n: usize) -> UniquePtr<BasisSet>;

        fn set_pure(basis: Pin<&mut BasisSet>, solid: bool);

        /// Number of [`Shell`]s in the basis.
        #[must_use]
        fn nshells(basis: &BasisSet) -> usize;
        fn nbf(basis: &BasisSet) -> usize;
        fn max_nprim(basis: &BasisSet) -> usize;
        fn max_l(basis: &BasisSet) -> usize;
        fn shell2bf(basis: &BasisSet) -> &CxxVector<usize>;

        fn at(basis: &BasisSet, i: usize) -> UniquePtr<Shell>;
    }
}
